{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "18c8e6e9fb190222",
   "metadata": {},
   "source": [
    "# Summary Statistics\n"
   ]
  },
  {
   "cell_type": "code",
   "id": "eb7259fa3d6f27ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:47.686732Z",
     "start_time": "2024-07-29T19:33:47.535990Z"
    }
   },
   "source": [
    "import re\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "import sqlalchemy as sa\n",
    "from sqlalchemy import text"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "code",
   "id": "5b55a1a4fe533c17",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:47.705274Z",
     "start_time": "2024-07-29T19:33:47.687975Z"
    }
   },
   "source": [
    "# find the current working directory\n",
    "import os\n",
    "import IPython\n",
    "\n",
    "notebook_path = IPython.get_ipython().getoutput('pwd')[0]\n",
    "\n",
    "data_dir = os.path.join(notebook_path, '..', 'data')\n",
    "\n",
    "if not os.path.exists(data_dir):\n",
    "    os.makedirs(data_dir)"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "id": "ef6ebb85b13db416",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:47.710731Z",
     "start_time": "2024-07-29T19:33:47.707091Z"
    }
   },
   "source": [
    "from datetime import datetime\n",
    "\n",
    "date = datetime.now().strftime(\"%Y-%m-%d\")\n",
    "\n",
    "# create dir for figures\n",
    "figures_dir = os.path.join(notebook_path, \"..\", \"figures\", \"analysis_5\", date)\n",
    "\n",
    "if not os.path.exists(figures_dir):\n",
    "    print(f\"Creating directory: {figures_dir}\")\n",
    "    os.makedirs(figures_dir)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating directory: /Users/ac4294/dev/2024/cpvt_database_analysis/notebooks/../figures/analysis_5/2024-07-29\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "bf3a4b9c79e6e81f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:47.715584Z",
     "start_time": "2024-07-29T19:33:47.712150Z"
    }
   },
   "source": [
    "sns.set_style(\"ticks\")\n",
    "\n",
    "DEFAULT_COLOR = sns.color_palette(\"husl\", 9)[6]"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "id": "342b04cebd62cf1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:47.721469Z",
     "start_time": "2024-07-29T19:33:47.717240Z"
    }
   },
   "source": [
    "def add_bar_labels(ax, fmt=\"{:.0f}\", ignore_zero=True):\n",
    "    for bar in ax.patches:\n",
    "\n",
    "        if bar.get_height() == 0 and ignore_zero:\n",
    "            continue\n",
    "\n",
    "        # Use ax.text() to add text over the bars\n",
    "        ax.text(bar.get_x() + bar.get_width() / 2, bar.get_height(),\n",
    "                fmt.format(bar.get_height()),\n",
    "                ha='center', va='bottom')\n",
    "\n",
    "\n",
    "def save_current_plot(name: str):\n",
    "    for fmt in [\"png\", \"pdf\", \"svg\"]:\n",
    "        plt.savefig(\n",
    "            os.path.join(figures_dir, name + '.' + fmt), dpi=300)"
   ],
   "outputs": [],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "id": "61fcda05384cb32b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.001032Z",
     "start_time": "2024-07-29T19:33:47.722847Z"
    }
   },
   "source": [
    "# read db connection from .env file\n",
    "from helpers.settings import get_settings\n",
    "\n",
    "settings = get_settings()"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading settings ...\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "id": "90363085adcc4eb0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.107230Z",
     "start_time": "2024-07-29T19:33:50.002322Z"
    }
   },
   "source": [
    "engine = sa.create_engine(settings.postgresql_dsn.__str__())"
   ],
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Information on Patients, Variants, Publications\n",
    "\n",
    "Basic information on how many patients, patients with CPVT, number of variants (in protein / genomic level), number of publications"
   ],
   "id": "e2790cd0756f00d3"
  },
  {
   "cell_type": "code",
   "id": "f54c8403cb7e0869",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.205818Z",
     "start_time": "2024-07-29T19:33:50.111862Z"
    }
   },
   "source": [
    "def qry(sql: str):\n",
    "    with engine.connect() as conn:\n",
    "        res = conn.execute(text(sql)).scalar()\n",
    "\n",
    "    return res\n",
    "\n",
    "\n",
    "tot_variants = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT variant_id)\n",
    "    FROM individual_variant\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of variants: {tot_variants} (TOTAL)\")\n",
    "\n",
    "num_cpvt_variants = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT variant_id) AS num_patients\n",
    "    FROM individual_variant\n",
    "    JOIN cpvt_patients_v cp\n",
    "    ON cp.individual_id = individual_variant.individual_id\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of variants in CPVT patients: {num_cpvt_variants}\")\n",
    "\n",
    "num_tot_patients = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(individual_id) AS num_patients\n",
    "    FROM individual\n",
    "\"\"\"\n",
    ")\n",
    "print(f\"Number of patients: {num_tot_patients} (TOTAL)\")\n",
    "\n",
    "num_cpvt_patients = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT individual_id) AS num_cpvt_patients\n",
    "    FROM cpvt_patients_v\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of CPVT patients: {num_cpvt_patients}\")\n",
    "\n",
    "num_pubs = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT p.publication_id)\n",
    "    FROM publication p\n",
    "    JOIN individual_to_publication itp \n",
    "    ON p.publication_id = itp.publication_id\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of publications: {num_pubs}\")\n",
    "\n",
    "hgvs_p_parsable = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.sequence_variant_id) AS num_protein_variants_hgvs_parsed\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    WHERE sv.p_hgvs_string IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of PROTEIN variant_ids with parsed HGVS: {hgvs_p_parsable} (TOTAL)\")\n",
    "\n",
    "hgvs_cpvt_p_parsable = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.sequence_variant_id) AS num_protein_variants_hgvs_parsed\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    JOIN cpvt_patients_v cp\n",
    "    ON cp.individual_id = iv.individual_id\n",
    "    WHERE sv.p_hgvs_string IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of PROTEIN variant_ids with parsed HGVS in CPVT patients: {hgvs_cpvt_p_parsable}\")\n",
    "\n",
    "hgvs_variants_total_patients = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.sequence_variant_id) AS num_hgvs_parsable\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    WHERE sv.sequence_variant_id IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of unique variant_ids (protein/dna) parsable: {hgvs_variants_total_patients} in (TOTAL)\")\n",
    "\n",
    "hgvs_any_parsable_cpvt = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.sequence_variant_id) AS num_hgvs_parsable\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    JOIN cpvt_patients_v cp\n",
    "    ON cp.individual_id = iv.individual_id\n",
    "    WHERE sv.sequence_variant_id IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of unique variant_ids (protein/dna) parsable: {hgvs_any_parsable_cpvt} in CPVT patients\")\n",
    "\n",
    "hgvs_only_hgvs_str = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.p_hgvs_string) AS num_unique_proteins\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    WHERE sv.sequence_variant_id IS NOT NULL\n",
    "    AND sv.p_hgvs_string IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of unique protein HGVS STRINGS: {hgvs_only_hgvs_str} (TOTAL)\")\n",
    "\n",
    "hgvs_only_hgvs_str_cpvt = qry(\n",
    "    \"\"\"\n",
    "    SELECT COUNT(DISTINCT sv.p_hgvs_string) AS num_unique_proteins\n",
    "    FROM individual_variant iv\n",
    "    JOIN variant v\n",
    "    ON iv.variant_id = v.variant_id\n",
    "    JOIN sequence_variant sv\n",
    "    ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "    JOIN cpvt_patients_v cp\n",
    "    ON cp.individual_id = iv.individual_id\n",
    "    WHERE sv.sequence_variant_id IS NOT NULL\n",
    "    AND sv.p_hgvs_string IS NOT NULL\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "print(f\"Number of unique protein HGVS STRINGS: {hgvs_only_hgvs_str_cpvt} in CPVT patients\")"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variants: 393 (TOTAL)\n",
      "Number of variants in CPVT patients: 284\n",
      "Number of patients: 1347 (TOTAL)\n",
      "Number of CPVT patients: 968\n",
      "Number of publications: 225\n",
      "Number of PROTEIN variant_ids with parsed HGVS: 382 (TOTAL)\n",
      "Number of PROTEIN variant_ids with parsed HGVS in CPVT patients: 275\n",
      "Number of unique variant_ids (protein/dna) parsable: 392 in (TOTAL)\n",
      "Number of unique variant_ids (protein/dna) parsable: 283 in CPVT patients\n",
      "Number of unique protein HGVS STRINGS: 381 (TOTAL)\n",
      "Number of unique protein HGVS STRINGS: 274 in CPVT patients\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.346847Z",
     "start_time": "2024-07-29T19:33:50.207234Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# get the list of variants not parsable\n",
    "variants_not_parsable = pd.read_sql(\"\"\"\n",
    "SELECT v.variant_id,\n",
    "    v.hgvs_string\n",
    "FROM variant v\n",
    "JOIN individual_variant iv\n",
    "ON v.variant_id = iv.variant_id\n",
    "WHERE v.sequence_variant_id IS NULL\n",
    "\"\"\", engine)\n",
    "variants_not_parsable"
   ],
   "id": "988956891a8f7e8f",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    variant_id      hgvs_string\n",
       "0         9225  exon 3 deletion\n",
       "1         9225  exon 3 deletion\n",
       "2         9225  exon 3 deletion\n",
       "3         9225  exon 3 deletion\n",
       "4         9225  exon 3 deletion\n",
       "5         9225  exon 3 deletion\n",
       "6         9225  exon 3 deletion\n",
       "7         9225  exon 3 deletion\n",
       "8         9225  exon 3 deletion\n",
       "9         9225  exon 3 deletion\n",
       "10        9225  exon 3 deletion\n",
       "11        9225  exon 3 deletion\n",
       "12        9225  exon 3 deletion\n",
       "13        9225  exon 3 deletion\n",
       "14        9225  exon 3 deletion"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>hgvs_string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9225</td>\n",
       "      <td>exon 3 deletion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.351123Z",
     "start_time": "2024-07-29T19:33:50.347969Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def add_pct(df: pd.DataFrame, cat_col: str, col: str = \"num_patients\"):\n",
    "    df[\"percentage\"] = df[col] / df[col].sum() * 100\n",
    "\n",
    "    df[\"percentage_exclude_nan\"] = df.dropna(subset=[cat_col])[col] / df.dropna(subset=[cat_col])[col].sum() * 100\n",
    "\n",
    "    return df"
   ],
   "id": "f6cbd5e71e5e3f3f",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Sex of patients",
   "id": "28e1145577f59cc2"
  },
  {
   "cell_type": "code",
   "id": "759ef76c21fc14f1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.372539Z",
     "start_time": "2024-07-29T19:33:50.352491Z"
    }
   },
   "source": [
    "# get number male, female, NAN\n",
    "sex_agg = pd.read_sql(\"\"\"\n",
    "SELECT s.value AS sex,\n",
    "    COUNT(DISTINCT individual_id) AS num_patients\n",
    "FROM individual i \n",
    "LEFT JOIN individual_sex s\n",
    "ON i.individual_sex_id = s.individual_sex_id\n",
    "GROUP BY s.value \n",
    "\"\"\", engine)\n",
    "add_pct(sex_agg, \"sex\")\n",
    "sex_agg"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      sex  num_patients  percentage  percentage_exclude_nan\n",
       "0  female           645   47.884187               54.987212\n",
       "1    male           528   39.198218               45.012788\n",
       "2    None           174   12.917595                     NaN"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>female</td>\n",
       "      <td>645</td>\n",
       "      <td>47.884187</td>\n",
       "      <td>54.987212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>male</td>\n",
       "      <td>528</td>\n",
       "      <td>39.198218</td>\n",
       "      <td>45.012788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>174</td>\n",
       "      <td>12.917595</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Only CPVT Patients",
   "id": "1b06de32cd9afd29"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.391358Z",
     "start_time": "2024-07-29T19:33:50.373819Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# only cpvt patients\n",
    "only_cpvt_sex_agg = pd.read_sql(\"\"\"\n",
    "SELECT s.value AS sex,\n",
    "    COUNT(DISTINCT i.individual_id) AS num_patients\n",
    "FROM cpvt_patients_v cpv \n",
    "LEFT JOIN individual i\n",
    "ON cpv.individual_id = i.individual_id\n",
    "LEFT JOIN individual_sex s\n",
    "ON i.individual_sex_id = s.individual_sex_id\n",
    "GROUP BY s.value\n",
    "\"\"\", engine)\n",
    "add_pct(only_cpvt_sex_agg, \"sex\")\n",
    "only_cpvt_sex_agg"
   ],
   "id": "f8c18f93fa04a333",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      sex  num_patients  percentage  percentage_exclude_nan\n",
       "0  female           464   47.933884               56.862745\n",
       "1    male           352   36.363636               43.137255\n",
       "2    None           152   15.702479                     NaN"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>female</td>\n",
       "      <td>464</td>\n",
       "      <td>47.933884</td>\n",
       "      <td>56.862745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>male</td>\n",
       "      <td>352</td>\n",
       "      <td>36.363636</td>\n",
       "      <td>43.137255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>152</td>\n",
       "      <td>15.702479</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Inheritance",
   "id": "3725fa171a277e9d"
  },
  {
   "cell_type": "code",
   "id": "690efdfd5f7592a7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.406897Z",
     "start_time": "2024-07-29T19:33:50.392799Z"
    }
   },
   "source": [
    "# mutation inheritance\n",
    "inheritance_agg = pd.read_sql(\"\"\"\n",
    "SELECT vi.variant_inheritance AS inheritance,\n",
    "    COUNT (DISTINCT iv.individual_id) AS num_patients\n",
    "FROM individual_variant iv\n",
    "JOIN variant_inheritance vi\n",
    "ON iv.variant_inheritance_id = vi.variant_inheritance_id\n",
    "GROUP BY variant_inheritance\n",
    "\"\"\", engine)\n",
    "add_pct(inheritance_agg, \"inheritance\")\n",
    "inheritance_agg"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   inheritance  num_patients  percentage  percentage_exclude_nan\n",
       "0    inherited           251   71.509972               71.509972\n",
       "1  spontaneous           100   28.490028               28.490028"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>inheritance</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>inherited</td>\n",
       "      <td>251</td>\n",
       "      <td>71.509972</td>\n",
       "      <td>71.509972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>spontaneous</td>\n",
       "      <td>100</td>\n",
       "      <td>28.490028</td>\n",
       "      <td>28.490028</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 13
  },
  {
   "cell_type": "code",
   "id": "f4be3182986d74e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.412199Z",
     "start_time": "2024-07-29T19:33:50.407986Z"
    }
   },
   "source": [
    "total_inheritance_records = inheritance_agg['num_patients'].sum()\n",
    "total_inheritance_records"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(351)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 14
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# HGVS Type\n",
    "\n",
    "By num unique variants (only protein level)"
   ],
   "id": "5429a1b43a04ee79"
  },
  {
   "cell_type": "code",
   "id": "8c55b4b79b3e56d9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.433821Z",
     "start_time": "2024-07-29T19:33:50.413616Z"
    }
   },
   "source": [
    "# get edit type of p variant\n",
    "p_edit_type = pd.read_sql(\"\"\"\n",
    "SELECT et.name AS edit_type,\n",
    "    COUNT(DISTINCT v.variant_id) AS num_variants\n",
    "FROM individual_variant iv\n",
    "JOIN variant v\n",
    "ON iv.variant_id = v.variant_id\n",
    "LEFT JOIN  sequence_variant sv\n",
    "ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "LEFT JOIN edit_type et\n",
    "ON sv.p_edit_type = et.edit_type_id\n",
    "GROUP BY name\n",
    "\"\"\", engine)\n",
    "# calculate pct\n",
    "add_pct(p_edit_type, \"edit_type\", \"num_variants\")\n",
    "p_edit_type"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      edit_type  num_variants  percentage  percentage_exclude_nan\n",
       "0      Deletion             5    1.272265                1.308901\n",
       "1    Frameshift             1    0.254453                0.261780\n",
       "2     Insertion             3    0.763359                0.785340\n",
       "3  Substitution           373   94.910941               97.643979\n",
       "4          None            11    2.798982                     NaN"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edit_type</th>\n",
       "      <th>num_variants</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Deletion</td>\n",
       "      <td>5</td>\n",
       "      <td>1.272265</td>\n",
       "      <td>1.308901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Frameshift</td>\n",
       "      <td>1</td>\n",
       "      <td>0.254453</td>\n",
       "      <td>0.261780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Insertion</td>\n",
       "      <td>3</td>\n",
       "      <td>0.763359</td>\n",
       "      <td>0.785340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Substitution</td>\n",
       "      <td>373</td>\n",
       "      <td>94.910941</td>\n",
       "      <td>97.643979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>11</td>\n",
       "      <td>2.798982</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "By patients",
   "id": "331d3359b3e470c6"
  },
  {
   "cell_type": "code",
   "id": "f2c33af9755c709c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.451377Z",
     "start_time": "2024-07-29T19:33:50.435087Z"
    }
   },
   "source": [
    "p_edit_type_by_patient = pd.read_sql(\"\"\"\n",
    "SELECT et.name AS edit_type,\n",
    "    COUNT(iv.individual_id) AS num_patients\n",
    "FROM individual_variant iv\n",
    "JOIN variant v\n",
    "ON iv.variant_id = v.variant_id\n",
    "LEFT JOIN  sequence_variant sv\n",
    "ON v.sequence_variant_id = sv.sequence_variant_id\n",
    "LEFT JOIN edit_type et\n",
    "ON sv.p_edit_type = et.edit_type_id\n",
    "WHERE name IS NOT NULL\n",
    "GROUP BY name\n",
    "\"\"\", engine)\n",
    "# calculate pct\n",
    "add_pct(p_edit_type_by_patient, \"edit_type\", \"num_patients\")\n",
    "p_edit_type_by_patient"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      edit_type  num_patients  percentage  percentage_exclude_nan\n",
       "0      Deletion            26    1.977186                1.977186\n",
       "1     Insertion             4    0.304183                0.304183\n",
       "2  Substitution          1282   97.490494               97.490494\n",
       "3    Frameshift             3    0.228137                0.228137"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edit_type</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Deletion</td>\n",
       "      <td>26</td>\n",
       "      <td>1.977186</td>\n",
       "      <td>1.977186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Insertion</td>\n",
       "      <td>4</td>\n",
       "      <td>0.304183</td>\n",
       "      <td>0.304183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Substitution</td>\n",
       "      <td>1282</td>\n",
       "      <td>97.490494</td>\n",
       "      <td>97.490494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Frameshift</td>\n",
       "      <td>3</td>\n",
       "      <td>0.228137</td>\n",
       "      <td>0.228137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Zygosity",
   "id": "a6ab3709ae54f39b"
  },
  {
   "cell_type": "code",
   "id": "c9dc7afb720197aa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.465217Z",
     "start_time": "2024-07-29T19:33:50.452562Z"
    }
   },
   "source": [
    "zygosity_agg = pd.read_sql(\"\"\"\n",
    "                           SELECT z.zygosity,\n",
    "                            COUNT(iv.individual_id) AS num_patients\n",
    "                            FROM individual_variant iv\n",
    "                            JOIN zygosity z\n",
    "                            ON iv.zygosity_id = z.zygosity_id\n",
    "                            GROUP BY z.zygosity\n",
    "                            \"\"\", engine)\n",
    "add_pct(zygosity_agg, \"zygosity\")\n",
    "zygosity_agg"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "       zygosity  num_patients  percentage  percentage_exclude_nan\n",
       "0  heterozygous           632   99.059561               99.059561\n",
       "1    homozygous             6    0.940439                0.940439"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>zygosity</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>heterozygous</td>\n",
       "      <td>632</td>\n",
       "      <td>99.059561</td>\n",
       "      <td>99.059561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>homozygous</td>\n",
       "      <td>6</td>\n",
       "      <td>0.940439</td>\n",
       "      <td>0.940439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 17
  },
  {
   "cell_type": "code",
   "id": "5ff2035e8685b71",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.470803Z",
     "start_time": "2024-07-29T19:33:50.466358Z"
    }
   },
   "source": [
    "total_zygosity_records = zygosity_agg['num_patients'].sum()\n",
    "total_zygosity_records"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(638)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 18
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Only CPVT Patients\n",
   "id": "bea471f3a7540a22"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.553131Z",
     "start_time": "2024-07-29T19:33:50.472229Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# conditions\n",
    "conditions_to_graph = [\n",
    "    \"Exercise/stress induced polymorphic ventricular tachycardia\",\n",
    "    \"Syncope exercise/stress induced\",\n",
    "    \"Exercise/stress induced sudden cardiac arrest\",\n",
    "    \"Baseline/resting electrocardiogram abnormality\",\n",
    "    \"Heart Structure Abnormality\",\n",
    "]\n",
    "\n",
    "\n",
    "def make_condition_qry(\n",
    "        condition: str\n",
    "):\n",
    "    stmt = text(\"\"\"\n",
    "    SELECT\n",
    "    c.condition,\n",
    "    ic.has_condition,\n",
    "    COUNT(ic.individual_id) AS num_patients\n",
    "    FROM individual_condition ic\n",
    "    JOIN condition c\n",
    "    ON c.condition_id = ic.condition_id\n",
    "    WHERE c.condition ILIKE :condition\n",
    "    AND ic.has_condition IS NOT NULL\n",
    "    AND ic.individual_id IN (\n",
    "        SELECT individual_id\n",
    "        FROM individual_condition ic2\n",
    "        JOIN condition c2\n",
    "        ON c2.condition_id = ic2.condition_id\n",
    "        WHERE c2.condition = 'Catecholaminergic polymorphic ventricular tachycardia 1'\n",
    "        AND ic2.has_condition = true\n",
    "    )   \n",
    "    GROUP BY c.condition, ic.has_condition\n",
    "    ORDER BY c.condition, ic.has_condition\n",
    "    \"\"\")\n",
    "\n",
    "    return stmt.bindparams(condition=condition)\n",
    "\n",
    "\n",
    "all_dfs = []\n",
    "\n",
    "for condition in conditions_to_graph:\n",
    "    df_condition = pd.read_sql(make_condition_qry(condition), engine)\n",
    "    add_pct(df_condition, \"has_condition\")\n",
    "\n",
    "    # print total in group\n",
    "    total_in_group = df_condition['num_patients'].sum()\n",
    "    print(f\"Total patients with {condition}: {total_in_group}\")\n",
    "\n",
    "    all_dfs.append(df_condition)\n",
    "\n",
    "# merge all dataframes\n",
    "all_conditions = pd.concat(all_dfs)\n",
    "all_conditions"
   ],
   "id": "b0b572d9ccfa2bab",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total patients with Exercise/stress induced polymorphic ventricular tachycardia: 323\n",
      "Total patients with Syncope exercise/stress induced: 271\n",
      "Total patients with Exercise/stress induced sudden cardiac arrest: 113\n",
      "Total patients with Baseline/resting electrocardiogram abnormality: 465\n",
      "Total patients with Heart Structure Abnormality: 494\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "                                           condition  has_condition  \\\n",
       "0  Exercise/stress induced polymorphic ventricula...          False   \n",
       "1  Exercise/stress induced polymorphic ventricula...           True   \n",
       "0                    Syncope exercise/stress induced          False   \n",
       "1                    Syncope exercise/stress induced           True   \n",
       "0      Exercise/stress induced sudden cardiac arrest          False   \n",
       "1      Exercise/stress induced sudden cardiac arrest           True   \n",
       "0     Baseline/resting electrocardiogram abnormality          False   \n",
       "1     Baseline/resting electrocardiogram abnormality           True   \n",
       "0                        Heart Structure Abnormality          False   \n",
       "1                        Heart Structure Abnormality           True   \n",
       "\n",
       "   num_patients  percentage  percentage_exclude_nan  \n",
       "0             7    2.167183                2.167183  \n",
       "1           316   97.832817               97.832817  \n",
       "0             9    3.321033                3.321033  \n",
       "1           262   96.678967               96.678967  \n",
       "0            16   14.159292               14.159292  \n",
       "1            97   85.840708               85.840708  \n",
       "0           388   83.440860               83.440860  \n",
       "1            77   16.559140               16.559140  \n",
       "0           449   90.890688               90.890688  \n",
       "1            45    9.109312                9.109312  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>has_condition</th>\n",
       "      <th>num_patients</th>\n",
       "      <th>percentage</th>\n",
       "      <th>percentage_exclude_nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exercise/stress induced polymorphic ventricula...</td>\n",
       "      <td>False</td>\n",
       "      <td>7</td>\n",
       "      <td>2.167183</td>\n",
       "      <td>2.167183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Exercise/stress induced polymorphic ventricula...</td>\n",
       "      <td>True</td>\n",
       "      <td>316</td>\n",
       "      <td>97.832817</td>\n",
       "      <td>97.832817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Syncope exercise/stress induced</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>3.321033</td>\n",
       "      <td>3.321033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Syncope exercise/stress induced</td>\n",
       "      <td>True</td>\n",
       "      <td>262</td>\n",
       "      <td>96.678967</td>\n",
       "      <td>96.678967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exercise/stress induced sudden cardiac arrest</td>\n",
       "      <td>False</td>\n",
       "      <td>16</td>\n",
       "      <td>14.159292</td>\n",
       "      <td>14.159292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Exercise/stress induced sudden cardiac arrest</td>\n",
       "      <td>True</td>\n",
       "      <td>97</td>\n",
       "      <td>85.840708</td>\n",
       "      <td>85.840708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Baseline/resting electrocardiogram abnormality</td>\n",
       "      <td>False</td>\n",
       "      <td>388</td>\n",
       "      <td>83.440860</td>\n",
       "      <td>83.440860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Baseline/resting electrocardiogram abnormality</td>\n",
       "      <td>True</td>\n",
       "      <td>77</td>\n",
       "      <td>16.559140</td>\n",
       "      <td>16.559140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Heart Structure Abnormality</td>\n",
       "      <td>False</td>\n",
       "      <td>449</td>\n",
       "      <td>90.890688</td>\n",
       "      <td>90.890688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Heart Structure Abnormality</td>\n",
       "      <td>True</td>\n",
       "      <td>45</td>\n",
       "      <td>9.109312</td>\n",
       "      <td>9.109312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Treatments\n",
    "\n",
    "Must `JOIN` with `cpvt_patients_v` to get only CPVT patients"
   ],
   "id": "c59ed6328399a933"
  },
  {
   "cell_type": "code",
   "id": "c6b7b4ba8b5847cb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.565972Z",
     "start_time": "2024-07-29T19:33:50.554251Z"
    }
   },
   "source": [
    "# get num patients with detailed treatment records\n",
    "num_patients = pd.read_sql(\"\"\"\n",
    "SELECT COUNT(DISTINCT patient_id) AS num_patients\n",
    "FROM treatment_record tr\n",
    "JOIN treatment t\n",
    "ON tr.treatment_id = t.treatment_id\n",
    "WHERE tr.patient_id IN (\n",
    "    SELECT individual_id\n",
    "    FROM cpvt_patients_v\n",
    ") AND\n",
    "tr.treatment_taken IS NOT NULL\n",
    "\"\"\", engine)\n",
    "num_patients"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   num_patients\n",
       "0           605"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.578845Z",
     "start_time": "2024-07-29T19:33:50.567745Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# beta blocker + any other treatment\n",
    "num_patients_bb = pd.read_sql(\"\"\"\n",
    "SELECT COUNT(DISTINCT patient_id) AS num_patients\n",
    "FROM treatment_record tr\n",
    "JOIN treatment t\n",
    "ON tr.treatment_id = t.treatment_id\n",
    "WHERE tr.patient_id IN (\n",
    "    SELECT individual_id\n",
    "    FROM cpvt_patients_v\n",
    ") AND\n",
    "tr.treatment_taken = TRUE AND\n",
    "treatment_name = 'Beta blocker'\n",
    "\"\"\", engine)\n",
    "num_patients_bb"
   ],
   "id": "8a0963f3f34d97d1",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   num_patients\n",
       "0           434"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.585443Z",
     "start_time": "2024-07-29T19:33:50.580770Z"
    }
   },
   "cell_type": "code",
   "source": "num_patients_bb['num_patients'] / num_patients['num_patients']",
   "id": "69c8fd554228092d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.717355\n",
       "Name: num_patients, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 22
  },
  {
   "cell_type": "code",
   "id": "7a5083f05d6e1c63",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.605718Z",
     "start_time": "2024-07-29T19:33:50.587514Z"
    }
   },
   "source": [
    "# get num patients with beta blocker + \n",
    "# Verapamil\n",
    "# Flecainide\n",
    "# Enalapril\n",
    "\n",
    "num_patients_poly_pharm = pd.read_sql(\"\"\"\n",
    "SELECT COUNT(DISTINCT patient_id) AS num_patients\n",
    "FROM treatment_record tr\n",
    "JOIN treatment t\n",
    "ON tr.treatment_id = t.treatment_id\n",
    "WHERE tr.patient_id IN (\n",
    "    SELECT individual_id\n",
    "    FROM cpvt_patients_v\n",
    ") AND\n",
    "tr.treatment_taken = TRUE AND \n",
    "tr.patient_id IN (\n",
    "    SELECT tr.patient_id\n",
    "    FROM treatment_record tr\n",
    "    JOIN treatment t\n",
    "    ON tr.treatment_id = t.treatment_id\n",
    "    WHERE t.treatment_name = 'Beta blocker'\n",
    "    AND tr.treatment_taken = TRUE\n",
    ")\n",
    "AND tr.patient_id IN (\n",
    "    SELECT tr.patient_id\n",
    "    FROM treatment_record tr\n",
    "    JOIN treatment t\n",
    "    ON tr.treatment_id = t.treatment_id\n",
    "    WHERE t.treatment_name IN ('Flecainide', 'Verapamil', 'Enalapril')\n",
    "    AND tr.treatment_taken = TRUE\n",
    ")\n",
    "\"\"\", engine)\n",
    "\n",
    "num_patients_poly_pharm"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   num_patients\n",
       "0           132"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 23
  },
  {
   "cell_type": "code",
   "id": "7564b73e4105059c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.611460Z",
     "start_time": "2024-07-29T19:33:50.607125Z"
    }
   },
   "source": [
    "# calculate percentage\n",
    "num_patients_poly_pharm['num_patients'] / num_patients['num_patients']"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.218182\n",
       "Name: num_patients, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 24
  },
  {
   "cell_type": "code",
   "id": "e69f724a8be15244",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.629833Z",
     "start_time": "2024-07-29T19:33:50.612859Z"
    }
   },
   "source": [
    "# invasive treatment\n",
    "# Left cardiac sympathetic denervation\n",
    "# Implantable cardioverter-defibrillator\n",
    "# Catheter ablation\n",
    "num_patients_invasive = pd.read_sql(\"\"\"\n",
    "SELECT COUNT(DISTINCT patient_id) AS num_patients\n",
    "FROM treatment_record tr\n",
    "JOIN treatment t\n",
    "ON tr.treatment_id = t.treatment_id\n",
    "WHERE tr.patient_id IN (\n",
    "    SELECT individual_id\n",
    "    FROM cpvt_patients_v\n",
    ") AND\n",
    "tr.treatment_taken = TRUE AND \n",
    "tr.patient_id IN (\n",
    "    SELECT tr.patient_id\n",
    "    FROM treatment_record tr\n",
    "    JOIN treatment t\n",
    "    ON tr.treatment_id = t.treatment_id\n",
    "    WHERE t.treatment_name = 'Beta blocker'\n",
    "    AND tr.treatment_taken = TRUE\n",
    ")\n",
    "AND tr.patient_id IN (\n",
    "    SELECT tr.patient_id\n",
    "    FROM treatment_record tr\n",
    "    JOIN treatment t\n",
    "    ON tr.treatment_id = t.treatment_id\n",
    "    WHERE t.treatment_name IN ('Left cardiac sympathetic denervation', 'Implantable cardioverter-defibrillator', 'Catheter ablation')\n",
    "    AND tr.treatment_taken = TRUE\n",
    ")\n",
    "\"\"\", engine)\n",
    "\n",
    "num_patients_invasive"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   num_patients\n",
       "0           130"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_patients</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 25
  },
  {
   "cell_type": "code",
   "id": "77f197de742b2a16",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.635174Z",
     "start_time": "2024-07-29T19:33:50.630917Z"
    }
   },
   "source": [
    "num_patients_invasive['num_patients'] / num_patients['num_patients']"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0.214876\n",
       "Name: num_patients, dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 26
  },
  {
   "cell_type": "code",
   "id": "94d93153f2bf9690",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-07-29T19:33:50.638203Z",
     "start_time": "2024-07-29T19:33:50.636535Z"
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": 26
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
